README.md
# trialMinRand

This is a barebone, no-frills R package that implements
a minimization randomization algorithm for clinical trials.
The package implements minimization as described in:

Pocock, Stuart J.; Simon, Richard:
__Sequential Treatment Assignment with Balancing for Prognostic Factors in the Controlled Clinical Trial.__
_Biometrics_, __31__(1):103-115, 1975. https://dx.doi.org/10.2307/2529712


## Example

This is a basic example which shows you how to solve a common problem:

```R
# Define patient data
d <- data.frame( F1=factor(c(1,2,2),levels=1:2),
  F2=factor(c(1,1,2),levels=1:2) )

# Define previous treatments
tr <- c("A","A")

# Define desired ratio for all possible treatments: A:B=2:1
tratio <- c(A=2,B=1)

# Call function to obtain next treatment. With probability 
# 0.7, this is going to be a treatment that minimizes imbalance.
# With probability 0.3, it is going to be a random treatment.
# Both imbalance and the distribution of random treatments are 
# defined using a 2:1 ratio for the treatments A and B.
next.tr <- assign.next.treatment( d, tratio, tr, p=0.3 )
```


ChangeLog
2017-03-23 Inge Wortel <ingewortel@gmail.com>

	* README.md : Updated example.

	* trialMinRand.R (simulate.trial) : added option 'method' which can be either 'min' (minimization) or 'rand' (random assignment). Merged with simulate.random.trial()

	* trailMinRand.R (simulate.random.trial) : removed (merged with simulate.trial)

	* trailMinRand.R (simulate.trial) : Added obligatory input argument 'strata.levels'; a named list with all possible values for the different factors in the patient.data dataframe. Ensures correct behaviour if columns of patient.data are not factors and not all strata are present in the initial data. Changed addition random new patients accordingly.

	* trialMinRand.R (simulate.trial) : For random treatment assignment, added probabilities according to treatment.ratios to the sampling of a new treatment. 	


2017-03-22 Inge Wortel <ingewortel@gmail.com>

	* ChangeLog : created

	* trailMinRand.R : updated documentation.

	* trailMinRand.R (simulate.trial) : Forced conversion of treatments vector to character to avoid problems with factor behaviour. Added treatment.ratios as obligatory input argument to make function compatible with the changes in imbalance() and assign.next.treatment() on 13/03. Changed .c.fac() to normal c().

	* trailMinRand.R (simulate.random.trial) : Likewise. Added treatment.ratios as an argument so the names can be used to sample a treatment from (instead of the factor levels, which are not necessarily there because treatments does not have to be a factor anymore). 

	* trailMinRand.R (total.imbalance) : Added obligatory argument treatment.ratios() to pass on to imbalance().



2017-03-13 Inge Wortel <ingewortel@gmail.com>

	* trialMinRand.R (assign.next.treatment): convert factors to character vector before calculating imbalance (avoid unexpected behaviour of factors resulting in a bug that the order of treatments in treatment.ratios influences result).

	* trialMinRand.R (assign.next.treatment): removed default for treatment.ratios. This forces the user to specify all possible treatments in treatment.ratios which prevents a situation where at treatment never gets assigned because it wasn't present in the input.
	
	* trialMinRand.R (imbalance) : likewise.

	* trialMinRand.R (assign.next.treatment): previous.treatment no longer has to be a factor, can be character or numeric vector (as long as all treatments in this vector correspond to the names in treatment.ratios).

	* trialMinRand.R (imbalance) : likewise

	* trialMinRand.R (.check.treatments) : now allows character/numeric vectors next to factors.

	* trialMinRand.R (.check.treatment.ratios) : added checks that 1) treatment.ratios is a vector, 2) treatment.ratios is numeric, 3) treatment.ratios is named.

	* trialMinRand.R (.check.treatment.names) : new function to check that all treatments in previous.treatments correspond to one of the names in treatment.ratios. If previous.treatments is a factor, the levels of this factor should all correspond to one of the names in treatment.ratios.

	* trialMinRand.R (.c.fac) : removed; no longer required now that factors are internally converted to character.

	* trialMinRand.R (imbalance) : bugfix - start with an empty table (0s for all treatments present in treatment.ratios) to prevent that treatments are missing from the table if they haven't occurred yet in previous.treatments. 

	* trialMinRand.R (imbalance) : updated documentation.
	
	* trialMinRand.R (assign.next.treatment) : likewise


	* test-imbalance.R : Added test to check input: 
		- if no input is specified
		- if input doesn't satisfy .check.treatments(), .check.treatment.ratios() or .check.treatment.names()

	* test-imbalance.R : Added test to check output: should be numeric of length one. This test should fail if imbalance ever returns NA (logical).

	* test-imbalance.R : Added tests for function calls that should be equivalent:
		- result should be the same for different encoding of treatment (factor, numeric or character)
		- result should stay the same if treatment.ratios is multiplied with a constant (eg 1:1 = 3:3).
		- result should stay the same if the order in the treatment vector changes.
		- result should stay the same if the order of treatments in treatment.ratios changes
		- For 1:1 ratio, it shouldn't matter which treatment is present less.

	* test-imbalance.R : Added tests for specific output values
		- balanced treatment vectors (factor, numeric or character) should always return 0
		- unbalanced vectors of any type should never return zero.

	* test-imbalance.R : Added test to check that function still works if not all treatments have been assigned yet.

	* test-assignment.R : File created. NOTE: all tests of assign.next.treatment() that are based on a value use p=0 to avoid randomness in treatment assignment. 

	* test-assignment.R : Added tests to check input; errors for wrong input:
		- if no input is specified	
		- if either patient.data or treatment.ratios is missing
		- if patient.data is not a dataframe of at least one row
		- if input doesn't satisfy the .check.treatments(), .check.treatment.ratios() 

	* test-assignment.R : Added test to check input compatibility:
		- errors returned if nrow(patient.data) != length(previous.treatments)
		- errors if input does not satisfy .check.treatment.names()

	* test-assignment.R : Added test to check the output format:
		- output should be character of length 1. (This test should fail if imbalance() ever returns NA.)

	* test-assignment.R : Added tests for function calls that should be equivalent:
		- result should stay the same if treatment.ratios is multiplied with a constant (50:100 vs 1:2)
		- result should stay the same if elements of treatment.ratios are in a different order
		- result should stay the same if previous patients are ordered differently
		  (reorder previous.treatments and all but the last row of patient.data in the same way).
		
	* test-assignment.R : Added tests for specific output values:

tests/testthat.R
library(testthat)
library(trialMinRand)

test_check("trialMinRand")


tests/testthat/test-imbalance.R
# Test for correct format of the input data:
# ---------------------

# Check that all required inputs are there
test_that("imbalance returns an error when not all required inputs are specified.",{
	expect_error(imbalance())
	expect_error(imbalance(c("A","B","B","A")))
	expect_error(imbalance(treatment.ratios=c(A=1,B=2)))
	expect_error(imbalance(c(A=1,B=2)))
})

# Check that inputs are of the right structure (if not, there should be erros from .check.treatments() and .check.treatment.ratios()
test_that("imbalance returns an error when inputs are not in the correct format.",{

	# Errors if treatments is not a vector 
		expect_error(imbalance(matrix(2,2,2),treatment.ratios=c(A=1,B=2)),
			"treatments must be a vector or a factor!")

	# Errors if treatment.ratios is unnamed, not a numeric vector 
	# or has non-integer/negative values:
		expect_error(imbalance(c("A","B","B","A"), matrix(1,3,3)),
			"treatment.ratios must be a vector!")
		expect_error(imbalance(c("A","B","B","A"), c(A="a",B="b")))
		expect_error(imbalance(c("A","B","B","A"), c(A="1",B="2")))
		expect_error(imbalance(c("A","B","B","A"), c(1,1)),
			"The vector treatment.ratios must be named after the treatments!")
		expect_error(imbalance(c("A","B","B","A"), c(A=1,B=0.25)),
			"Please use integers to specify treatment ratios!")
		expect_error(imbalance(c("A","B","B","A"), c(A=3.25,B=1)),
			"Please use integers to specify treatment ratios!")
		expect_error(imbalance(c("A","B","B","A"), c(A=1,B=-2)),
			"Each treatment ratio must be positive!")
})

# Check that elements in 'treatments' correspond to the names of 'treatment.ratios'
test_that("all elements in 'treatments' are present in 'treatment.ratios' ",{

	# Different treatment names in 'treatments' vs 'treatment.ratios'
		expect_error(imbalance(c(1,2,2,4),c(A=1,B=1,C=1)),
			"Not all treatments can be found in treatment.ratios.")
	
	# There is a treatment that is absent from 'treatment.ratios'
		expect_error(imbalance(c(1,2,2,4),c("1"=1,"2"=1)),
			"Not all treatments can be found in treatment.ratios.")
		expect_error(imbalance(c("banana","apple","banana","apple","pear"),c(apple=1,banana=1)),
			"Not all treatments can be found in treatment.ratios.")

	# Treatments is a factor, but the levels do not correspond to names of treatment.ratios
		expect_error(imbalance(factor(c(1,2,2,1),levels=c(1,2,3)),c("1"=1,"2"=2)))
})


# Test for correct output format
# ---------------------
test_that("Output is numeric and has length 1",{
	expect_equal(is.numeric(imbalance(c("apple"),c(apple=1,banana=1))),TRUE)
	expect_equal(is.numeric(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),
		treatment.ratios=c(A=1,B=1))),TRUE)
	expect_length(imbalance(c("placebo","placebo","X","placebo","X"),c(placebo=1,X=2)),1)
	expect_length(imbalance(c(3,4,3,3,4,5),c("3"=1,"5"=1,"4"=1)),1)
})


# Test for function calls that should be equivalent
# ---------------------


#Encoding type of treaments as factor() character() or numeric() doesn't matter
test_that("Encoding type of treatments ( - eg as numeric(), factor() or character() ) does not affect the output.",{
	expect_equal(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)),
		imbalance(factor(c(2,1,1,2),levels=c(2,1)),treatment.ratios=c("1"=1,"2"=1)))
	expect_equal(imbalance(c(1,2,3,2), treatment.ratios=c("1"=1,"2"=2,"3"=3)),
		imbalance(c("apple","pear","banana","pear"), treatment.ratios=c(apple=1,pear=2,banana=3)))	
})


#Output not changed when treatment.ratios is multiplied with a constant value:
test_that("multiplying all values in treatment.ratios with a constant does not alter output",{
	expect_equal(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)),
		imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=50,B=50)))
	expect_equal(imbalance(c(2,1,1,2),treatment.ratios=c("1"=1,"2"=2)),
		imbalance(c(2,1,1,2), treatment.ratios=c("1"=4,"2"=8)))
	expect_equal(imbalance(c("pear","apple","pear"), treatment.ratios=c("pear"=2,"apple"=3)),
		imbalance(c("pear","apple","pear"), treatment.ratios=c("pear"=40,"apple"=60)))
	expect_equal(imbalance(c("pear"), treatment.ratios=c("pear"=4,"apple"=1)),
		imbalance(c("pear"), treatment.ratios=c("pear"=40,"apple"=10)))
})

#Output not changed when treatments have a different order within 'treatments' vector
test_that("Changing the order of the treatments within the treatments vector does not alter output",{
	expect_equal(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)),
		imbalance(factor(c("A","A","B","B"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)))
	expect_equal(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)),
		imbalance(factor(c("A","B","A","B"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)))
	expect_equal(imbalance(c("pear","apple","pear"), treatment.ratios=c("pear"=2,"apple"=3)),
		imbalance(c("pear","pear","apple"), treatment.ratios=c("pear"=2,"apple"=3)))
	expect_equal(imbalance(c(1,2,1,2,1),treatment.ratios=c("1"=1,"2"=5)),
		imbalance(c(2,2,1,1,1),treatment.ratios=c("1"=1,"2"=5)))
})

#Output not changed when order of treatments is changed within treatment.ratios
test_that("Changing the order of the treatments within treatment.ratios does not alter output",{
	expect_equal(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=2)),
		imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(B=2,A=1)))
	expect_equal(imbalance(c("pear","apple","pear"), treatment.ratios=c("pear"=2,"apple"=3)),
		imbalance(c("pear","apple","pear"), treatment.ratios=c("apple"=3,"pear"=2)))
	expect_equal(imbalance(c(1,2,1,2,1),treatment.ratios=c("1"=1,"2"=5)),
		imbalance(c(1,2,1,2,1),treatment.ratios=c("2"=5,"1"=1)))
	expect_equal(imbalance(c("placebo","placebo","X","placebo","X"),c(placebo=1,X=2)),
		imbalance(c("placebo","placebo","X","placebo","X"),c(X=2,placebo=1)))
})

#In a 1:1 ratio, it doesn't matter which treatment is present less
test_that("In a 1:1 ratio, it it doesn't matter which treatment is present less.",{
	expect_equal(imbalance(factor(c("A","B","B"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)),
		imbalance(factor(c("A","A","B"),levels=c("A","B")),treatment.ratios=c(B=1,A=1)))
	expect_equal(imbalance(c("pear"), treatment.ratios=c("pear"=3,"apple"=3)),
		imbalance(c("apple"), treatment.ratios=c("pear"=3,"apple"=3)))
	expect_equal(imbalance(c(1,2,2,2), treatment.ratios=c("2"=50,"1"=50)),
		imbalance(c(1,2,1,1), treatment.ratios=c("2"=50,"1"=50)))
})


# Test for specific output values
# ---------------------

# Function imbalance should return an imbalance of 0 when the vector is balanced. 
test_that("imbalance measure is 0 for balanced vectors",{
	expect_equal(imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=1)),0)
	expect_equal(imbalance(factor(c("A","B","A","B","A","A"),levels=c("A","B")),treatment.ratios=c(A=2,B=1)),0)
	expect_equal(imbalance(factor(c(1,2,2,1,2,2,2,2,1),levels=c(2,1)),treatment.ratios=c('1'=1,'2'=2)),0)
	expect_equal(imbalance(c("placebo","X","X","placebo","X"),treatment.ratios=c(placebo=2,X=3)),0)
})

# Imbalance is not 0 for unbalanced vectors
test_that("imbalance measures for unbalanced vectors",{
	expect_equal(imbalance(factor( c("A","B","B") ,levels=c("A","B") ), treatment.ratios=c(A=1,B=1)),1)
	expect_equal(imbalance(factor( c("B","B","A"), levels=c("A","B") ), treatment.ratios=c(A=1,B=1)),1)
	expect_equal(imbalance(factor( c("A","B"), levels=c("B","A") ), treatment.ratios=c(A=2,B=1)),1)
	expect_equal(imbalance(factor( c("A","B"), levels=c("A","B") ), treatment.ratios=c(A=3,B=4)),1)
	expect_equal(imbalance(factor( c("A","B"), levels=c("B","A") ), treatment.ratios=c(A=1,B=5)),4)
	expect_equal(imbalance(factor( c("A","B","B"), levels=c("A","B","C") ), treatment.ratios=c(A=2,B=3,C=5)),20)
})


##OTHER:
# ---------------------

# Not all the treatments in the treatment.ratios have to be present in 'treatments'. 
test_that("Not all treatments need to be present in the treatments vector",{
		expect_equal(imbalance(c("apple"),c(apple=1,banana=1)),1)
		expect_equal(imbalance(c("apple"),c(banana=1,apple=1)),1)
		expect_equal(imbalance(c("banana"),c(banana=1,apple=1)),1)
		expect_equal(imbalance(factor(c(1,2,2,1),levels=c(1,2,3)),c("1"=1,"2"=1,"3"=1)),2)
		expect_equal(imbalance(factor(c(1,2,2,1),levels=c(1,2,3)),c("2"=1,"1"=1,"3"=1)),2)
		expect_equal(imbalance(c(1,2,2,1),c("2"=1,"1"=1,"3"=1)),2)
})


R/trialMinRand.R
#' Treatment Imbalance
#' 
#' Provides a measure of treatment imbalance with respect to a desired ratio of
#' treatments.
#'
#' @param treatments a factor or a character/numeric vector containing a sequence of treatments.
#' @param treatment.ratios a named vector containing the desired treatment ratios.
#' Treatment.ratios should at least contain all unique elements of the treatments vector.
#'
#' @details
#' A treatment vector has imbalance 0 if the treatments occur in the desired ratio.
#' Otherwise, the imbalance is positive.
#' 
#' @examples
#' # A balanced treatment vector
#' imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=1,B=1))
#' # Still balanced if ratio is given as 50:50 instead of 1:1
#' imbalance(factor(c("A","B","B","A"),levels=c("A","B")),treatment.ratios=c(A=50,B=50))
#' # Treatments can also be provided as a numeric vector
#' imbalance(c(1,2,2,1),treatment.ratios=c("1"=1,"2"=1))
#' # Treatments can be provided as a character vector
#' imbalance(c("A","B","B","A"), treatment.ratios=c(A=1,B=1))
#' 
#' @export
imbalance <- function( treatments, treatment.ratios ) {

	#Check input	
	.check.treatments( treatments )
	.check.treatment.ratios( treatment.ratios )
	.check.treatment.names( treatments, treatment.ratios )

	#Count how often each treatment occurs in the current vector.
	#Add the treatments not in there to the table with a 'zero'.
	xt <- setNames(rep(0,length(names(treatment.ratios))),names(treatment.ratios))
	xt[names(table( treatments ))] <- table( treatments )

	#Assign each treatment a weight such that multiplying the elements of treatment.ratios with the
	#elements of treatment.weights yields an equal, integer number for all treatments. To get
	#treatment.weights that are also integer, this number can be the smallest common multiple (scm)
	#of the elements in treatment.ratios. Dividing the scm by the treatment.ratios vector then yields the
	#vector with weights:
	treatment.weights <- .scm(treatment.ratios)/treatment.ratios[names(xt)]

	#Now multiply treatment counts with treatment weights. The difference of the range yields the (weighted)
	#imbalance:
	weighted.xt <- xt*treatment.weights
	diff(range(weighted.xt))
}

# The greatest common divisor of a vector of integer numbers.
.gcd <- function( x ){

	#Calculate gcd with "Euclidian algorithm"
	if( length(x) < 2 ){
		return(x)
	}
	if( length(x) > 2 ){
		return( .gcd( c( x[1], .gcd( x[-1] ) ) ) )
	}

	x <- sort(x)
	xd <- diff(x)
	if( xd==0 ){
		return(x[1])
	}
	.gcd( c(x[1], xd ) )
}

# The smallest common multiple of a vector of integer numbers.
.scm <- function( x ){
	if( length(x) < 2 ){
		return(x)
	}
	if( length(x) > 2 ){
		return( .scm( c( x[1], .scm( x[-1] ) ) ) )
	}
	return( prod(x)  / .gcd(x) )
}

# Format check for treatment ratios:
.check.treatment.ratios <- function(x){
	if( !is.vector(x) ){
		stop("treatment.ratios must be a vector!")
	}
	if( !is.numeric(x) ){
		stop("treatment.ratios must be a (named) numeric vector!")
	}
	if( !all( unname(x) ==  as.integer(x) ) ){
		stop("Please use integers to specify treatment ratios! (E.g., c(A=2,B=1) instead of c(A=1,B=0.5))")
	}
	if( any( x <= 0 ) ){
		stop("Each treatment ratio must be positive!")
	}

	if( is.null(names(x)) ){
		stop("The vector treatment.ratios must be named after the treatments! (E.g., c(A=2,B=1) instead of c(1,2))")

	}
}

# Format check for the actual treatments.
.check.treatments <- function(x){
	if( !is.vector(x) & !is.factor(x) ){
		stop("treatments must be a vector or a factor!")
	}
}

# Last check: names in treatment.ratios must correspond to the actual treatment names. 
.check.treatment.names <- function(treatments,treatment.ratios){

	tnames <- names(treatment.ratios)
	if(any(!is.element(treatments,tnames))){
		stop("Not all treatments can be found in treatment.ratios. Please check that treatments are spelled correctly and that treatment.ratios is complete.")
	}
	if(is.factor(treatments)){
		if(any(!is.element(levels(treatments),tnames))){
			stop("Treatment is a factor, but levels do not correspond to the names in treatment.ratios.")
		}
	}	

}



#' @importFrom stats runif setNames
#' @importFrom utils tail
NULL

#' Assign Next Treatment for Patient
#'
#' This function assigns the next treatment of a patient,
#' given the previous treatments.
#'
#' @param patient.data a data frame containing the strata for each patient.
#' The last row of this data frame is assumed to contain the strata for the
#' new patient, who will receive the next treatment. Strata can be given
#' as a factor, or as columns that can be interpreted as a factor.
#' 
#' @param previous.treatments a factor or numeric/character vector containing 
#' the previous treatments. 'previous.treatments' should have an element for
#' each row in patient.data except the last (which represents the current patient
#' for whom no treatment has been determined yet). 
#'
#' @param treatment.ratios a named vector containing the desired ratios
#' in which treatments are to be given.
#'
#' @param p probability with which a random treatment will be chosen 
#' (with the desired ratio) instead of choosing the treatment that minimizes
#' imbalance.
#'
#' @examples
#' # Define patient data
#' d <- data.frame( F1=factor(c(1,2,2),levels=1:2),
#'   F2=factor(c(1,1,2),levels=1:2) )
#' # Define previous treatments
#' tr <- factor( c("A","A"), levels=c("A","B") )
#' # Call function to obtain next treatment
#' next.tr <- assign.next.treatment( d, c(A=2,B=1), tr )
#'
#' @export
assign.next.treatment <- function(
	patient.data, treatment.ratios, previous.treatments=character(0), p=0.3 ){
	
	# CHECK INPUT:
	# patient.data must be a dataframe.
	if (class(patient.data)!="data.frame"){
		stop("patient.data must be a dataframe with at least one row!")
	}

	if ( nrow(patient.data) == 0 ){
		stop("patient.data must be a dataframe with at least one row!")
	}

	# treatment.ratios should be a named vector of non-negative integers.
	.check.treatment.ratios( treatment.ratios )

	# previous.treatments should be a vector of length nrow(patient.data) -1 and
	# should only contain treatments specified in the treatment.ratios vector.
	.check.treatments( previous.treatments )
	previous.treatments <- as.character(previous.treatments)

	if( length(previous.treatments) != nrow(patient.data)-1 ){
		stop(paste("The number of previous treatments must equal the number of patients - 1! Input data has",length(previous.treatments),"previous treatments and",nrow(patient.data),"patients."))
	}
	.check.treatment.names( previous.treatments, treatment.ratios )
	

	# CALCULATION
	treatments <- names(treatment.ratios)
	N <- length(treatments)

	# Compute the total imbalances for each treatment.
	factors <- names(patient.data)
	imbalances <- sapply( treatments, function(tr){ 
		sum( sapply( factors, function(f){
			f.values <- patient.data[[f]]
			imbalance( c(previous.treatments,tr)[f.values==tail(f.values,1)],
				treatment.ratios)
			} ) ) 
	} )

	if( runif(1) > p && diff(range(imbalances))>0 ){
		# Sort treatments in increasing order by imbalance, randomly breaking ties.
		treatments <- treatments[order(imbalances, sample(N))]
		# Return treatment with smallest imbalance.
		r <- treatments[1]
	} else {
		# Pick treatment at random, using biased coin to attain desired
		# treatment ratio.
		r <- sample( rep(treatments,treatment.ratios[treatments]), 1 )
	}
	r
}

simulate.trial <- function( initial.data, treatment.ratios, initial.treatments, N, strata.levels, method="min",... ){
	d <- initial.data 		# a dataframe
	tr <- initial.treatments 	# factor or numeric/character vector

	# Convert all to character vector to avoid problems with factor behaviour:
	tr <- as.character( tr )

	# Patients come in in random order until all have been treated. 
	while( nrow(d) < N ){
		# The new patient has a random combination of strata
		# All possible levels for each stratum are stored in the named list 'strata.levels'
		# (names must correspond to initial.data colnames)
		d <- rbind(d, sapply(colnames(d), function(x) sample(strata.levels[[x]],1)))

		# Assign the treatment according to the chosen method:
		if( method=="min"){
			#use assign.next.treatment
			tr <- c( tr, assign.next.treatment(d, treatment.ratios, tr,...) )

		} else if (method == "rand"){
		# Assign one of the treatments randomly with probability according to treatment.ratio
		tr <- c( tr, sample(names(treatment.ratios),1,prob=treatment.ratios) )

		} else {
			stop("Unknown method! Please pick 'min' or 'rand' or keep default.")
		}
			
	}
	# output both the renewed data and the renewed treatment vector.
	list( patient.data=d, treatments=tr )
}

total.imbalance <- function( patient.data, treatments, treatment.ratios, ... ){
	# For each factor f...
	apply( patient.data, 2, function(f){
		# ...Calculate the sum of the imbalances (imbalance for each treatment)
		sum( by( treatments, f, function(x) imbalance(x, treatment.ratios, ...) ) )
	})
}

